File: benchmarks/xlang/src/kmeans.R
n <- 200000; k <- 10; iters <- 15
xs <- numeric(n); ys <- numeric(n)
seed <- 42
for (i in 1:n) {
  seed <- (seed * 48271) %% 2147483647
  xs[i] <- 10.0 * seed / 2147483647.0
  seed <- (seed * 48271) %% 2147483647
  ys[i] <- 10.0 * seed / 2147483647.0
}
t0 <- proc.time()[["elapsed"]]
stride <- n %/% k
cx <- numeric(k); cy <- numeric(k)
for (c in 1:k) { cx[c] <- xs[(c - 1) * stride + 1]; cy[c] <- ys[(c - 1) * stride + 1] }
asgn <- integer(n)
for (it in 1:iters) {
  for (i in 1:n) {
    best <- 1; bd <- 1000000.0
    for (c in 1:k) {
      dx <- xs[i] - cx[c]; dy <- ys[i] - cy[c]
      d <- dx * dx + dy * dy
      if (d < bd) { bd <- d; best <- c }
    }
    asgn[i] <- best
  }
  sx <- numeric(k); sy <- numeric(k); ct <- integer(k)
  for (i in 1:n) {
    c <- asgn[i]
    sx[c] <- sx[c] + xs[i]; sy[c] <- sy[c] + ys[i]; ct[c] <- ct[c] + 1L
  }
  for (c in 1:k) if (ct[c] > 0) { cx[c] <- sx[c] / ct[c]; cy[c] <- sy[c] / ct[c] }
}
sizes <- integer(k)
for (i in 1:n) sizes[asgn[i]] <- sizes[asgn[i]] + 1L
cat("CHECK", max(sizes), "\n")
cat("MS", as.integer((proc.time()[["elapsed"]] - t0) * 1000), "\n")

File: benchmarks/xlang/src/kmeans.cpp
#include <cstdio>
#include <chrono>
#include <vector>
#include <algorithm>
int main() {
    const int n = 200000, k = 10, iters = 15;
    std::vector<double> xs(n), ys(n);
    long seed = 42;
    for (int i = 0; i < n; i++) {
        seed = seed * 48271 % 2147483647;
        xs[i] = 10.0 * seed / 2147483647.0;
        seed = seed * 48271 % 2147483647;
        ys[i] = 10.0 * seed / 2147483647.0;
    }
    auto t0 = std::chrono::steady_clock::now();
    int stride = n / k;
    std::vector<double> cx(k), cy(k);
    for (int c = 0; c < k; c++) { cx[c] = xs[c * stride]; cy[c] = ys[c * stride]; }
    std::vector<int> assign(n, 0);
    for (int it = 0; it < iters; it++) {
        for (int i = 0; i < n; i++) {
            int best = 0; double bd = 1000000.0;
            for (int c = 0; c < k; c++) {
                double dx = xs[i] - cx[c], dy = ys[i] - cy[c];
                double d = dx * dx + dy * dy;
                if (d < bd) { bd = d; best = c; }
            }
            assign[i] = best;
        }
        std::vector<double> sx(k, 0), sy(k, 0);
        std::vector<int> ct(k, 0);
        for (int i = 0; i < n; i++) {
            int c = assign[i];
            sx[c] += xs[i]; sy[c] += ys[i]; ct[c]++;
        }
        for (int c = 0; c < k; c++) if (ct[c] > 0) { cx[c] = sx[c] / ct[c]; cy[c] = sy[c] / ct[c]; }
    }
    std::vector<int> sizes(k, 0);
    for (int i = 0; i < n; i++) sizes[assign[i]]++;
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::steady_clock::now() - t0).count();
    printf("CHECK %d\nMS %lld\n", *std::max_element(sizes.begin(), sizes.end()), (long long)ms);
}
